### ISSUE TITLE
Crashing when adding to transform stream

### ISSUE THREAD

**Issue by 1mike12** [leeoniya/uDSV#10]
I'm no stream expert, but given the examples, it seemed like it would be super easy to create a transform stream with your module. 

I am able to verify that it works fine when running like the example in the readme, but when I port it into a custom Transform stream, it ends super early, doesn't hit the writeable, and I get no errors. There's a good chance this isn't a problem with uDSV, but it's strange, given your apis, this should work. 

```ts

// lines = 16_182_971 , 100% finished in 13.1 seconds = 1_235_341 rows per second
// 597.5 MB  13.1s   45.5MB/s
const streamedCSV = async () => {
  const fileSize = await fs.promises.stat(filePath).then((stats) => stats.size)
  const stream = fs.createReadStream(filePath).pipe(new ProgressBarStream(fileSize))

  let parser: Parser;
  for await (const buffer of stream) {
    const strChunk = buffer.toString();
    parser ??= initParser(inferSchema(strChunk));
    parser.chunk<string[]>(strChunk, parser.typedArrs, (batch) => {
      const x = batch // reaches here fine
    });
  }
  parser!.end();
}

const transform = async () => {
  const fileSize = await fs.promises.stat(fileWithHeader).then((stats) => stats.size)
  return new Promise<void>((resolve, reject) => {
    pipeline(
      fs.createReadStream(fileWithHeader),
      new SimpleUDSVTransform(),
      new Writable({
        write(chunk, encoding, callback) {
          const x = chunk // never reaches here
          callback()
        }
      }),
      (err) => {
        if (err) {
          reject(err) // never reaches here
        } else {
          resolve() // never reaches here
        }
      }
    )
  })
}
(async () => {
  await streamedCSV() //works
  await transform() //breaks
})()
```

**Comment by leeoniya** [comment]
> I'm no stream expert

heh, me neither :sweat_smile:. uDSV's api is fully synchronous, so there shouldn't be any surprises there.

does this reproduce with something you can attach here and i can debug (that hopefully isnt 16M records)?

**Comment by 1mike12** [comment]
yeah for sure, I just created a example repo you can use to run everything. just run example.ts

https://github.com/1mike12/usdv-stream-example

But regarding the api, I think the way to get it to work with streams is to use the chunk function, which seems to be async right now. I imagine it should be able to be synchronous though too though, since there shouldn't be a reason for the parser to not be able to give us what it's already got so far in a sync way?


```ts
this.parser.chunk<string[]>(strChunk, this.parser.stringArrs, (parsedData) => {
      this.push(parsedData);
      callback()
    });
```

But either way, whether its sync or async, it really doesn't matter to a Transform stream, which is why I'm so confused why it just stops and ends so soon

**Comment by leeoniya** [comment]
just heads up, it might be week or so before i can get to testing this.

**Comment by broncha** [comment]
@1mike12 I just implemented a transformstream with uDSV yesterday, and I ran into this as well.
The cause here seems to be `parser.chunk` does not parse the chunk that is provided, but instead queues it for later and the first chunk is only parsed on the second request.

```
chunk(csvStr, parse = stringArrs, cb = accum) {
	streamParse ??= parse;
	streamCb    ??= cb;

	if (streamState === 1) {
		streamParse(prevUnparsed + pendChunk, streamCb);
		streamChunkNum++;
	}

	pendChunk = csvStr;
	streamState = 1;
},
```

Heres my initial implementation, that I expected to work, but it didnt

```
class CSVTransformStream extends TransformStream {
  private parser?: Parser
  private schema?: Schema

  constructor(typeDeep = false) {
    super({
      transform: (chunk, controller) => {
        if (this.parser === undefined) {
          this.schema = inferSchema(chunk)
          this.schema.cols.find((c) => c.name === 'code').type = 's'
          this.parser = initParser(this.schema)
        }

        return new Promise((resolve) => {
          this.parser?.chunk(chunk, typeDeep ? this.parser.typedDeep : this.parser.stringArrs, (parsed) => {
            controller.enqueue(parsed)
            resolve()
          })
        })
      }
    })
  }
}
```

the callback for the first chunk is never called. so the source stream never sends the following chunks. 

Heres is the implementation that works

```
class CSVTransformStream extends TransformStream {
  private parser?: Parser
  private schema?: Schema

  constructor(typeDeep = false) {
    super({
      transform: (chunk, controller) => {
        if (this.parser === undefined) {
          this.schema = inferSchema(chunk)
          this.schema.cols.find((c) => c.name === 'code').type = 's'
          this.parser = initParser(this.schema)
        }

        return new Promise((resolve) => {
          this.parser?.chunk(chunk, typeDeep ? this.parser.typedDeep : this.parser.stringArrs, (parsed) => {
            controller.enqueue(parsed)
          })
          resolve()
        })
      },
      flush: () => {
        this.parser?.end()
      },
    })
  }
}
```

with the resolve outside the callback, it instructs the source stream to send following chunks, and as the chunks are parsed, the result are written to the writable side of transform. and the `flush` parses the final chunk.

**Comment by 1mike12** [comment]
@broncha that's awesome, thanks for figuring it out. I tried your fix in my example code and it got it working

https://github.com/1mike12/usdv-stream-example/blob/master/src/streams/SimpleUDSVTransform.ts

**Comment by leeoniya** [comment]
:pray: thanks @broncha!

@1mike12  now that this is figured out, wanna do a PR to add it to the streaming examples?

https://github.com/leeoniya/uDSV?tab=readme-ov-file#incremental--streaming

**Comment by 1mike12** [comment]
I remember first encountering this problem while inserting 300mb of data, I don't believe being under the threshold was it, but I can't remember 100% lol

**Comment by leeoniya** [comment]
diving back into the code i now remember why it's structured the way it is.

since the callback fires in chunkSize batches, the parser needs to know if what it's currently parsing can be treated as EOF so that if a chunk is smaller than chunkSize, it can call the callback without waiting for the next chunk to hit the chunkSize threshold. since streaming implementations do not provide the EOF signal as part of the chunk callback, but as a separate event, we have no way of knowing whether the _current_ chunk is the one with EOF - we only know this *after* the final chunk has been seen already. so the actual parsing has to be delayed by one chunk so we can parse it with an EOF flag.

i think the solution proposed here is the correct one.

**Comment by broncha** [comment]
@leeoniya the problem with this approach is, it kind of defeats the purpose of streaming. The readable side is notified to send more chunks immediately and all those chunks are buffered at uDSV (now since uDSV is fast, this may not be an issue at all). 
I am not dealing with large data, so its not an issue for me.

**Comment by leeoniya** [comment]
there is only a single chunk in the buffer at a time, so i dont think that will ever be an issue, you just pay the cost of having one chunk in memory, which should be insignificant since most of these chunks are measured in double digit KB.

the more unfortunate issue is if you want to have a live display of a stream of intermittent events, the parser will always be a chunk behind, which may be a few thousand records. im not sure if this is a real problem since i expect events to emit full records or batches without having the last record be partial. so effectively each event/batch will have an EOF and people can simply use uDSV's non incremental parser apis. nevertheless it would be good not to have this caveat, so i'll see if i can rethink the eof/batching situation.

**Comment by leeoniya** [comment]
this WIP [no-Chunksize branch](https://github.com/leeoniya/uDSV/tree/no-chunkSize) is showing some promise, even +10% perf improvement so far. the API will likely change a bit.

**Comment by leeoniya** [comment]
@1mike12 @broncha 

i've finished the work on the above branch and merged into `main`. the `.chunk()` api now behaves as you would expect without buffering 🎉 

an example has been added to the README and there is now a test for this below. please try it before i do the v0.7 release.

you can change your `package.json` dependency to `"udsv": "leeoniya/uDSV#ac6b47e0cc8dc9056d203c500c9431e340fe6cc3"`

https://github.com/leeoniya/uDSV/blob/ac6b47e0cc8dc9056d203c500c9431e340fe6cc3/test/parse.spec.mjs#L1088-L1140
